### ISSUE TITLE
git-gui: automatically select staged file before writing the commit msg

### ISSUE THREAD

**Issue by birgersp** [git-for-windows/git#2341]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

Hi all,

[I've raised the following issue in the git mailing list](https://public-inbox.org/git/CAGr--=+SNO7GuHH-dE_ZnrJDCa8tv8EA5LMrVGwsMTRpxhzEZA@mail.gmail.com/T/#m33da122fbcdfb46131af783e8bf5cf3e2033c4b6), already. So far 2 people have replied, but they didn't really voice their opinion whether this is a good idea or not. So I'd like to hear your thoughts

(copy-paste from the mail)

I personally prefer to have the changes I am about to commit visible
in the diff view, while I write my commit message. So usually I do
this:
1. Stage the file(s) I've been working on.
2. Select a file I just staged, so I can see the changes in the diff widget.
3. Jump to the "Commit Message" widget to type up my commit message.
Basically, I would like to be able to skip step 2. When the user
stages the last file in the "Unstaged Changes" widget, no file is
selected and the diff view becomes blank. When this is the case, I
would prefer that git-gui automatically selects one of the staged
files and shows it in the diff widget before I type up my commit
message. Naturally, this automatic selection should **only** happen
when the user chooses focus the "Commit Message" widget.

I propose:
(When the user focuses the "Commit Message" widget, if no file is
currently selected (i.e. diff widget shows no text))
automatically select the first file listed in the "Staged Changes"
widget so the changes of that file show up in the diff widget.

Thoughts?

**Comment by PhilipOakley** [comment]
[sorry for not replying on list]
I do end up doing that same window dance. 

I'm not sure if the auto swap 'focus' for the diff is 'right' for trying to understand what's happening for new user/humans (automated software can be bad - see MAX-737 for clues;-)

Maybe it needs a config setting `gui.commitShowsLastDiff` that can be set to true when one wants to try this, and the default can be flipped after a few cycles if it proves to be easy to understand and use.

However: it will need a proper config variable name. If you/we can't create a short config var name then it means its hard to explain, which means folk won't understand..

- gui.lastDiffFlipsFocus ?  When set, and the last file with unstaged changes is staged, the focus will follow that last file to staged files window, with that file selected" (insert the correct terminology)

**Comment by birgersp** [comment]
I can really relate to not wanting "automatic" software. My reasoning why this is an "okay" case is that a file will be auto-selected **only** when the diff view is blank. I don't think anyone prefers to see a blank diff view when writing their commit msg, so I feel that the user will enjoy that the changes appear in the diff view.

I'm all for it when it comes to config settings. I'd propose `gui.autoFocusStaged` as a variable name.

**Comment by dscho** [comment]
> I can really relate to not wanting "automatic" software. My reasoning why this is an "okay" case is that a file will be auto-selected **only** when the diff view is blank. I don't think anyone prefers to see a blank diff view when writing their commit msg, so I feel that the user will enjoy that the changes appear in the diff view.

FWIW I find this argument very convincing.

> I'm all for it when it comes to config settings. I'd propose `gui.autoFocusStaged` as a variable name.

That seems a fine name to me!

**Comment by dscho** [comment]
Closing as stale.
